File: cal/calsockethandler.go
// Copyright 2019 PayPal Inc.
//
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cal

import (
	"encoding/binary"
	"errors"
	"net"
	"syscall"
	"time"
	//"log"
)

const (
	//
	// max size of msg body.
	//
	calMessageBodyMaxSize = 4096
	handlerCtrlMsgNewRoot = "calHandlerCtrlmsgNewRoot"
)

type clientSocketStatus int

const (
	statusConnInProgress = iota
	statusConnected
	statusNotConnected
)

/**
 * cal socket handler is default. parent interface stays with default.
 */
type handler interface {
	init(_config *calConfig, _msgchann <-chan string) error
	run()
}

/**
 * gateway to caldaemon
 */
type socketHandler struct {

	//
	// readonly channel for arriving calmessages
	//
	mMsgReadChann <-chan string
	//
	// a message placeholder repopulated during each request
	//
	mMsg *clientMessage
	//
	// ring buffer to hold outgoing bytes (not thread safe)
	//
	mRingBuf *ringBuffer
	//
	//
	//
	mSocket *clientSocket
	//
	// a serialized calclientmsg for label. first msg to caldaemon after each reconnect
	//
	mLabel     []byte
	mLabelSize int
	//
	// indication of message loss to let next roottxn to reconnect
	//
	mRootTxnLossyFlag bool
	mConfig           *calConfig
}

/**
 * message between calclient and caldaemon
 */
type clientMessage struct {
	//
	// three header fields (stored as is, but going out in network byte order big endian)
	//
	mThreadID    uint32
	mConnectTime uint32
	mMsgLen      uint32

	mBody []byte
	mZero []byte // go has no memset. use this slice to clear 0 in one shot
}

/**
 * connection to caldaemon
 */
type clientSocket struct {
	mLastConnAttemptTime int
	mConnTime            int
	mStatus              clientSocketStatus
	mCaldAddress         string
	mConn                net.Conn
}

/********************************************************
 ********************************************************
 *               clientMessage
 ********************************************************
 */
func (sh *clientMessage) init(_label []byte, _enableTG bool) {
	sh.mBody = make([]byte, calMessageBodyMaxSize, calMessageBodyMaxSize)
	sh.mZero = make([]byte, calMessageBodyMaxSize, calMessageBodyMaxSize)
	sh.resetMessageHeader()

	//
	// add thread(group)id as the last four bytes of the msg. endianess really does not
	// matter since reading forward or backward produces the same uniqeness.
	//
	// setting threadid to 0 in callabelmsg force all calmsgs from the same process in
	// the same swimminglane in callogview, even when those calmsgs carry different
	// threadids in their calmsgheaders.
	//
	// occworker has a c++ calclient sending a callabelmsg with threadid==0. if their
	// msg arrives before proxy's callabelmsg having a threadid=0x1111, we cannot see
	// any proxy calmsg getting displayed (apparently, caldaemon rejects calmsgs having
	// a different threadid if the pool is started with threadid=0. in this case, we
	// need to set callabelmsg in proxy to match threadid==0. but, that would put
	// all calmsgs from proxy under the same swimminglane in callogview
	//
	// alternatively, we could make sure proxy fires off callabelmsg first. that way
	// calmsgs with different threadid in proxy will be spread into different
	// swimminglanes. at the same time, calmsg from different occworker processes would
	// also get their own separated lanes.
	//
	if _enableTG {
		_label = append(_label, 0, 0, 0, 1)
	} else {
		_label = append(_label, 0, 0, 0, 0)
	}
	sh.constructCalMessage(_label)
}

func (sh *clientMessage) resetMessageHeader() {
	//
	// @TODO messages having different thread id go to different vertical lanes in calview
	//
	sh.mThreadID = 0
	sh.mConnectTime = uint32(time.Now().Unix())
}

func (sh *clientMessage) toBytes() []byte {
	serialized := make([]byte, sh.getLength())
	binary.BigEndian.PutUint32(serialized[0:4], sh.mThreadID)
	binary.BigEndian.PutUint32(serialized[4:8], sh.mConnectTime)
	binary.BigEndian.PutUint32(serialized[8:12], sh.mMsgLen)
	copy(serialized[12:], sh.mBody[:sh.mMsgLen])
	return serialized
}

func (sh *clientMessage) getLength() uint32 {
	return sh.mMsgLen + 12 // 12 bytes header
}

func (sh *clientMessage) constructCalMessage(_buffer []byte) {
	/* *  This gets called for every single CAL message, so it
	 *  needs a little optimization.  There may not be a
	 *  trailing NUL on the end of the string, but it
	 *  wouldn't get copied anyway.  We get the string
	 *  length, ensure that it ends with at least one \r\n
	 *  (although it could end up ending with two), and
	 *  then use memcpy (fast!) to ship it out.
	 */
	copy(sh.mBody, sh.mZero)

	//
	// last four bytes is thread id
	//
	size := len(_buffer)
	sh.mThreadID = binary.LittleEndian.Uint32(_buffer[size-4:])
	_buffer = _buffer[0 : size-4]

	if len(_buffer) > calMessageBodyMaxSize {
		_buffer = _buffer[:calMessageBodyMaxSize]
	}
	copy(sh.mBody, _buffer)

	sh.mBody[calMessageBodyMaxSize-2] = '\r'
	sh.mBody[calMessageBodyMaxSize-1] = '\n'

	sh.mMsgLen = uint32(len(_buffer))
	//log.Println("constructcalmsg ", _buffer, sh.mMsgLen)
}

/********************************************************
 ********************************************************
 *               socketHandler
 ********************************************************
 *
 * Init is called by CalClient in a synchronized block
 */
func (sh *socketHandler) init(_config *calConfig, _msgchann <-chan string) error {
	if _config == nil {
		return errors.New("passing nil calconfig to calsockethandler")
	}
	sh.mConfig = _config

	sh.mRingBuf = &ringBuffer{}
	sh.mRingBuf.Init(_config.getRingBufferSize())

	sh.mMsg = &clientMessage{}
	sh.mMsg.init([]byte(_config.getLabel()), _config.enableTG)
	//
	// keep a serialized label message for each reconnection
	//
	sh.mLabelSize = int(sh.mMsg.getLength())
	sh.mLabel = sh.mMsg.toBytes()

	sh.mSocket = &clientSocket{}
	sh.mSocket.Init(_config.host, _config.port, int(_config.connTimeSec))
	sh.mSocket.establishCaldConnection()

	sh.mMsgReadChann = _msgchann

	//
	// put label on the ringbuf and let the next client write to flush it to caldaemon
	//
	sh.mRingBuf.WriteData(sh.mLabel, sh.mLabelSize)
	sh.mRootTxnLossyFlag = false

	return nil
}

/**
 * ringbuffer is not made thread safe. instead, we use a channel to ensure single
 * access to ringbuffer and clientsocket. calling other functions like addDataToRingBuffer
 * directly would cause race condition having undefined behavior(s)
 */
func (sh *socketHandler) run() {
	for {
		calmsg, ok := <-sh.mMsgReadChann
		if !ok {
			//
			// @TODO calclient closed channel, clean up memory and exit
			//
			return
		}

		switch calmsg {
		case handlerCtrlMsgNewRoot:
			sh.handleNewRootTransaction()
		default:
			sh.completeConnectionToCaldaemon()
			sh.addDataToRingBuffer(calmsg)
			sh.flush()
		}
	}
}

/**
 * none of these functions should be called from outside socketHandler
 */
func (sh *socketHandler) completeConnectionToCaldaemon() {
	if sh.mSocket.mStatus == statusNotConnected {
		sh.handleDisconnect()
	} else if sh.mSocket.mStatus == statusConnInProgress {
		sh.handleConnectionInProgress()
	}
}

func (sh *socketHandler) addDataToRingBuffer(_msg string) {
	sh.mMsg.constructCalMessage([]byte(_msg))

	if !sh.mRingBuf.WriteData(sh.mMsg.toBytes(), int(sh.mMsg.getLength())) {
		sh.mRootTxnLossyFlag = true
	}
}

func (sh *socketHandler) flush() {
	if sh.mSocket.mStatus != statusConnected {
		return
	}
	var size = sh.mRingBuf.mUsed
	buffer := make([]byte, size)
	sh.mRingBuf.CopyData(buffer, size)

	var dataSent = sh.mSocket.SendData(buffer, size)
	if dataSent < 0 {
		//m_logger->write_trace_message (CAL_LOG_WARNING, errno, "Closing connection: failed to send %d bytes. %d", size, errno);
		sh.closeConnection()
		sh.mRootTxnLossyFlag = true
	} else {
		sh.mRingBuf.RemoveData(dataSent)
	}
}

func (sh *socketHandler) handleDisconnect() {
	if sh.mSocket.establishCaldConnection() {
		sh.mMsg.resetMessageHeader()
	}
}

func (sh *socketHandler) handleConnectionInProgress() {
	/* This would fail in the following cases
	1) caldaemon down
	2) time out during connection progress
	In both of the cases it make sense to close connection */

	if !sh.mSocket.connectToCaldaemon() {
		sh.closeConnection()
	}
}

func (sh *socketHandler) closeConnection() {
	sh.mSocket.closeSocket()
	sh.mRingBuf.Clear()

	//
	// Always fill the RB with the label, after discarding its data
	//
	sh.mRingBuf.WriteData(sh.mLabel, sh.mLabelSize)
}

func (sh *socketHandler) handleNewRootTransaction() {
	if sh.mRootTxnLossyFlag {
		sh.flush()
		sh.closeConnection()
		sh.mRootTxnLossyFlag = false
		sh.completeConnectionToCaldaemon()
	}
}

/********************************************************
 ********************************************************
 *               clientSocket
 ********************************************************
 */
func (sh *clientSocket) Init(_host string, _port string, _connTime int) {
	sh.mLastConnAttemptTime = 0
	sh.mStatus = statusNotConnected
	sh.mConnTime = _connTime
	sh.mCaldAddress = _host + ":" + _port
}

func (sh *clientSocket) hasConnTimeoutExpired() bool {
	if int(time.Now().Unix()) >= (sh.mLastConnAttemptTime + sh.mConnTime) {
		return true
	}
	return false
}

func (sh *clientSocket) establishCaldConnection() bool {
	if !sh.hasConnTimeoutExpired() {
		return false
	}

	sh.mLastConnAttemptTime = int(time.Now().Unix())
	if !sh.connectToCaldaemon() {
		sh.closeSocket()
		return false
	}
	return true
}

func (sh *clientSocket) connectToCaldaemon() bool {
	//
	// connection timeout at 500 ms going after caldaemon on the same box.
	//
	var err error
	sh.mConn, err = net.DialTimeout("tcp", sh.mCaldAddress,
		time.Duration(500*time.Millisecond))
	if err != nil {
		return sh.checkSocketConnectErrors(err)
	}
	sh.mStatus = statusConnected
	return true
}

func (sh *clientSocket) checkSocketConnectErrors(_err error) bool {
	var rt = false
	//
	// @TODO test these error conditions on stage
	//
	switch t := _err.(type) {
	case syscall.Errno:
		switch t {
		case syscall.EINPROGRESS:
			sh.mStatus = statusConnInProgress
			rt = true
		case syscall.EISCONN:
			sh.mStatus = statusConnected
			rt = true
		case syscall.EALREADY:
			if !sh.hasConnTimeoutExpired() {
				sh.mStatus = statusConnInProgress
				rt = true
			}
		default:
		}
		//
		// all other type of errors return false. none of these are tested.
		//
		//case net.Error:
		//	if t.Timeout()
		//case *net.OpError:
		//	if opError.Op == "dial"/unknowhost
		//	if opError.Op == "read"/connrefused
	}
	return rt
}

func (sh *clientSocket) SendData(_msg []byte, _size int) int {
	if sh.mStatus != statusConnected {
		return 0
	}

	sh.mConn.SetWriteDeadline(time.Now().Add(time.Duration(500 * time.Millisecond)))
	sent, err := sh.mConn.Write(_msg)
	//log.Println("senddata ", _msg, sent, string(_msg[12:]))

	if (sent <= 0) && (err != nil) {
		switch t := err.(type) {
		case syscall.Errno:
			switch t {
			case syscall.EAGAIN:
				return 0
			default:
				return -1
			}
		default:
			return -1
		}
	}
	return sent
}

func (sh *clientSocket) closeSocket() {
	if sh.mStatus != statusNotConnected {
		if sh.mConn != nil {
			sh.mConn.Close()
		}
		sh.mStatus = statusNotConnected
	}
}

File: worker/cppworker/cal/CalSocketHandler.cpp

#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <errno.h>
#include <string.h>

#include "CalSocketHandler.h"
#include "CalRingBuffer.h"
#include "CalLog.h"
#include "CalConfig.h"
#include "CalClient.h"

CalSocketHandler::CalSocketHandler (CalConfig* _config, CalLog* _logger)
  :CalHandler(_logger),
  m_msg(_config->get_label()),
  m_label((const char*)&m_msg,m_msg.get_length()),
  m_client_socket(_config,_logger),
  m_ringbuffer(_config->get_ring_buffer_size(),_logger),
  m_root_txn_lossy_flag (false)
{
	add_label_to_ring_buffer ();
}

CalSocketHandler::~CalSocketHandler ()
{
}

CalSocketHandler::CalClientMessage::CalClientMessage(const std::string & _label)
{
	set_connection_time_in_message();
	construct_cal_message(_label);
}

void CalSocketHandler::CalClientMessage::construct_cal_message(const std::string& _buffer)
{
	/* *  This gets called for every single CAL message, so it
	 *  needs a little optimization.  There may not be a
	 *  trailing NUL on the end of the string, but it
	 *  wouldn't get copied anyway.  We get the string
	 *  length, ensure that it ends with at least one \r\n
	 *  (although it could end up ending with two), and
	 *  then use memcpy (fast!) to ship it out.
	 */
	memset(m_body,0,sizeof(m_body));

	int len = _buffer.length();
	if (len > CAL_MESSAGE_BUFFER_SIZE) 
	{
		len = CAL_MESSAGE_BUFFER_SIZE;
	}
	
	memcpy(m_body, _buffer.c_str(), len);
	m_body[CAL_MESSAGE_BUFFER_SIZE-2] = '\r';
	m_body[CAL_MESSAGE_BUFFER_SIZE-1] = '\n';
	m_header.m_msg_len = htonl (len);
}

unsigned int CalSocketHandler::CalClientMessage::get_length () const
{
	return ntohl(m_header.m_msg_len)+sizeof(Header);
}

void CalSocketHandler::CalClientMessage::set_connection_time_in_message()
{
	m_header.m_thread_id = 0;
	m_header.m_connect_time = htonl (time(NULL));
}

void CalSocketHandler::add_label_to_ring_buffer()
{
	m_ringbuffer.write_data(m_label.c_str(), m_label.length());
}

void CalSocketHandler::write_data(const std::string& _data)
{
	m_logger->write_trace_message(CAL_LOG_VERBOSE, 0, "%s",_data.c_str());

	complete_connection_to_caldaemon();
	add_data_to_ring_buffer(_data);
	flush();
}

void CalSocketHandler::complete_connection_to_caldaemon()
{
	if(m_client_socket.get_status() == NOT_CONNECTED)
	{
		handle_disconnect();
	}
	else
	if(m_client_socket.get_status() == CONNECTION_IN_PROGRESS)
	{
		handle_connection_in_progress();
	}
}

void CalSocketHandler::add_data_to_ring_buffer(const std::string& _data)
{
	m_msg.construct_cal_message(_data); 
	char* msg = (char*) &m_msg;
	int size =  m_msg.get_length();

	if(!m_ringbuffer.write_data(msg,size))
	{
		set_root_txn_lossy_flag(true);
	}
}

void CalSocketHandler::flush()
{
	if(m_client_socket.get_status() != CONNECTED)
		return;

	int size = m_ringbuffer.used_capacity();
	char buffer[size];
	m_ringbuffer.copy_data(buffer,size);

	int data_sent = m_client_socket.send_data(buffer,size);
	if(data_sent < 0)
	{
		m_logger->write_trace_message (CAL_LOG_WARNING, errno, "Closing connection: failed to send %d bytes. %d", size, errno);
		close_connection();
		set_root_txn_lossy_flag(true);
	}
	else
	{
		m_ringbuffer.remove_data(data_sent);
	}
}

void CalSocketHandler::handle_disconnect()
{
	if(m_client_socket.establish_cald_connection())
	{
		m_msg.set_connection_time_in_message();
	}
}

void CalSocketHandler::handle_connection_in_progress()
{
	/* This would fail in the following cases 
	1) caldaemon down
	2) time out during connection progress 
	In both of the cases it make sense to close connection */

	if(!m_client_socket.connect_to_caldaemon())
	{
		close_connection();
	}
}

void CalSocketHandler::close_connection ()
{
	m_client_socket.close_socket();
	discard_ring_buffer_contents();

	//Always fill the RB with the label, after discarding its data
	add_label_to_ring_buffer();
}

void CalSocketHandler::discard_ring_buffer_contents()
{
	m_ringbuffer.clear();
}

void CalSocketHandler:: handle_new_root_transaction()
{
	if (m_root_txn_lossy_flag)
	{
		flush();
		close_connection();

		set_root_txn_lossy_flag(false);

		complete_connection_to_caldaemon();
	}
}

void CalSocketHandler::set_root_txn_lossy_flag(bool _value)
{
	
	m_root_txn_lossy_flag = _value;
}

////////////////////////////////////
//
// CalClientSocket
//
////////////////////////////////////

 /*
 Give an initial value which is less than the current time, so that calclient always goes in connect mode during initialization 
 */

CalSocketHandler::CalClientSocket::CalClientSocket(CalConfig* _config, CalLog* _logger)
:m_last_connection_attempt_time(0),
 m_logger(_logger),
 m_config(_config),
 m_status(NOT_CONNECTED)
{
	establish_cald_connection();
}

CalSocketHandler::CalClientSocket::~CalClientSocket()
{
	close_socket();
}

void CalSocketHandler:: CalClientSocket::set_status(CalClientSocketStatus _status)
{
	m_status = _status;
}

void CalSocketHandler:: CalClientSocket::print_current_state() const
{
	switch(m_status)
	{
	case CONNECTED:
		m_logger->write_trace_message(CAL_LOG_VERBOSE, 0, "Changed status to CONNECTED");
		break;
	case NOT_CONNECTED:
		m_logger->write_trace_message(CAL_LOG_VERBOSE, 0, "Changed status to NOT_CONNECTED");
		break;
	case CONNECTION_IN_PROGRESS:
		m_logger->write_trace_message(CAL_LOG_VERBOSE, 0, "Changed status to CONNECTION_IN_PROGRESS");
		break;
	default:
		m_logger->write_trace_message(CAL_LOG_VERBOSE, 0, "UNKNOWN-STATE");
		break;
	}
}

CalSocketHandler::CalClientSocketStatus CalSocketHandler::CalClientSocket::get_status() const
{
	return m_status;
}

bool CalSocketHandler::CalClientSocket::has_reconnect_timeout_expired() const 
{
	if(has_timeout_expired())
	{
		return true;
	}
	return false;
}

bool CalSocketHandler::CalClientSocket::has_connection_timeout_expired() const 
{
	if(has_timeout_expired())
	{
		return true;
	}
	return false;
}

bool CalSocketHandler::CalClientSocket::has_timeout_expired() const 
{
	if (time(NULL) >= (m_last_connection_attempt_time+ m_config->get_socket_connect_time()))
	{
		return true;
	}
	return false;
}

bool CalSocketHandler::CalClientSocket::establish_cald_connection()
{
	if(!has_reconnect_timeout_expired()) 
		return false;

	// create socket
	if(!create_socket())
		return false;

	if(!set_socket_non_blocking())
	{
		close_socket();
		return false;
	}

	m_last_connection_attempt_time  = time(NULL);
	if(!connect_to_caldaemon())
	{
		close_socket();
		return false;
	}
	return true;
}

bool CalSocketHandler::CalClientSocket::create_socket()
{
	int s = ::socket (PF_INET, SOCK_STREAM, 0);
	if (s < 0)
	{
		return false;
	}
	m_socket = s;
	return true;
}

bool CalSocketHandler::CalClientSocket::set_socket_non_blocking()
{
	int flags = fcntl(m_socket, F_GETFL, 0);
	if (flags == -1)
	{
		return false;
	}
	fcntl(m_socket, F_SETFL, flags | O_NONBLOCK);
	return true;
}

bool CalSocketHandler::CalClientSocket::connect_to_caldaemon()
{
 	sockaddr_in addr = m_config->get_siteview_addr();
	if (::connect (m_socket, (struct sockaddr *) &addr, sizeof (addr)) < 0)
	{
		return check_socket_connect_errors();
	}

	set_status(CONNECTED);
	return true;
}

bool CalSocketHandler::CalClientSocket::check_socket_connect_errors()
{
	bool rt = false;
	switch(errno)
	{
	case EINPROGRESS:
		set_status(CONNECTION_IN_PROGRESS);
		rt = true;
		break;
	case EISCONN: 
		set_status(CONNECTED);
		rt = true;
		break;
	case EALREADY:
		if(!has_connection_timeout_expired())
		{
			set_status(CONNECTION_IN_PROGRESS);
			rt = true;
		}
		break;
	default:
		break;
	}
	return rt;
}

int CalSocketHandler::CalClientSocket::send_data(const char *msg, int size)
{
	if(get_status() != CONNECTED)
		return 0;

	int sent = ::send (m_socket, msg, size, MSG_NOSIGNAL);
	if( sent < 0)
	{
		int rt;
		switch(errno)
		{
		case EAGAIN:
			rt = 0;
			break;
		default:
			rt = -1;
			break;
		}
		return rt;
	}

	m_logger->write_trace_message (CAL_LOG_VERBOSE, 0, "Sent attempted %d bytes ,Sent %d bytes", size,sent);
	return sent;
}

void CalSocketHandler::CalClientSocket:: close_socket()
{
	if(get_status() != NOT_CONNECTED)
	{
		::close (m_socket);
		set_status(NOT_CONNECTED);
	}
}
